From b32ef1172046e4b3e9f496c3e06c363a12f3cbf4 Mon Sep 17 00:00:00 2001
From: Niall Pemberton <niallp@apache.org>
Date: Mon, 4 Feb 2008 08:55:29 +0000
Subject: [PATCH] Fix for BEANUTILS-302 - NPE in ArrayConverter when converting
 a string with underscores to a string array - reported by Martin Bartlett

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/beanutils/trunk@618207 13f79535-47bb-0310-9956-ffa450edef68
---
 .../beanutils/converters/ArrayConverter.java  |  8 +++---
 .../converters/ArrayConverterTestCase.java    | 27 +++++++++++++++++++
 2 files changed, 32 insertions(+), 3 deletions(-)

diff --git a/src/java/org/apache/commons/beanutils/converters/ArrayConverter.java b/src/java/org/apache/commons/beanutils/converters/ArrayConverter.java
index 900b82ada..da19ca231 100644
--- a/src/java/org/apache/commons/beanutils/converters/ArrayConverter.java
+++ b/src/java/org/apache/commons/beanutils/converters/ArrayConverter.java
@@ -434,10 +434,12 @@ private List parseElements(Class type, String value) {
             while (true) {
                 int ttype = st.nextToken();
                 if ((ttype == StreamTokenizer.TT_WORD) || (ttype > 0)) {
-                    if (list == null) {
-                        list = new ArrayList();
+                    if (st.sval != null) {
+                        if (list == null) {
+                            list = new ArrayList();
+                        }
+                        list.add(st.sval);
                     }
-                    list.add(st.sval.trim());
                 } else if (ttype == StreamTokenizer.TT_EOF) {
                     break;
                 } else {
diff --git a/src/test/org/apache/commons/beanutils/converters/ArrayConverterTestCase.java b/src/test/org/apache/commons/beanutils/converters/ArrayConverterTestCase.java
index 8867f7ce6..20ec30bdc 100644
--- a/src/test/org/apache/commons/beanutils/converters/ArrayConverterTestCase.java
+++ b/src/test/org/apache/commons/beanutils/converters/ArrayConverterTestCase.java
@@ -360,6 +360,33 @@ public void testErrors() {
         }
     }
 
+    /**
+     * Test for BEANUTILS-302 - throwing a NPE when underscore used
+     */
+    public void testUnderscore_BEANUTILS_302() {
+        String value = "first_value,second_value";
+        ArrayConverter converter = new ArrayConverter(String[].class, new StringConverter());
+
+        // test underscore not allowed (the default)
+        String[] result = (String[])converter.convert(String[].class, value);
+        assertNotNull("result.null", result);
+        assertEquals("result.length", 4, result.length);
+        assertEquals("result[0]", "first", result[0]);
+        assertEquals("result[1]", "value", result[1]);
+        assertEquals("result[2]", "second", result[2]);
+        assertEquals("result[3]", "value", result[3]);
+
+        // configure the converter to allow underscore
+        converter.setAllowedChars(new char[] {'.', '-', '_'});
+
+        // test underscore allowed
+        result = (String[])converter.convert(String[].class, value);
+        assertNotNull("result.null", result);
+        assertEquals("result.length", 2, result.length);
+        assertEquals("result[0]", "first_value", result[0]);
+        assertEquals("result[1]", "second_value", result[1]);
+    }
+
     /**
      * Check that two arrays are the same.
      * @param msg Test prefix msg
